File: src/main/java/study/codingtest/leetcode/sliding-window/LC_424.kt
fun characterReplacement(s: String, k: Int): Int {
  var left = 0
  val counts: MutableMap<Char, Int> = mutableMapOf()

  for(right in 1..s.length) {
    counts[s[right - 1]] = counts.getOrDefault(s[right - 1], 0) + 1
    val maxCharCount = counts.values.maxOrNull() ?: 0

    if (right - left - maxCharCount > k) {
      counts[s[left]] = counts.getOrDefault(s[left], 0) - 1
      left++
    }
  }
  return s.length - left
}

File: src/main/java/study/codingtest/leetcode/sliding-window/LC_424.ts
const characterReplacement = (s: string, k: number): number => {
  let left = 0;
  const counts: Map<string, number> = new Map();

  for (let right = 1; right <= s.length; right++) {
    counts.set(s[right - 1], (counts.get(s[right - 1]) || 0) + 1);
    const maxCharCount = Math.max(...Array.from(counts.values()));

    if (right - left - maxCharCount > k) {
      counts.set(s[left], (counts.get(s[left]) || 0) - 1);
      left++;
    }
  }
  return s.length - left;
};
